src/ripple/basics/algorithm.h

```diff
@@ -115,6 +116,45 @@ remove_if_intersect_or_match(
     return first1;
 }
 
+// Alternative implementation of `std::set_difference`
+// that takes a 3-way comparison function
+// and thus accepts input iterators with different value types.
+template <class InputIt1, class InputIt2, class OutputIt, class Compare>
+OutputIt
+set_diff(
+    InputIt1 first1,
+    InputIt1 last1,
+    InputIt2 first2,
+    InputIt2 last2,
+    OutputIt d_first,
+    Compare comp)
+{
+    while (first1 != last1)
+    {
+        if (first2 == last2)
+            return std::copy(first1, last1, d_first);
+
+        auto c = comp(*first1, *first2);
+        if (c < 0)
+            *d_first++ = *first1++;
+        else
+        {
+            if (c == 0)
+                ++first1;
+            ++first2;
+        }
+    }
+    return d_first;
+}
+
+/** Analog to std::move. */
+template <typename T>
+T
+copy(T const& t)
```

Comment:
how is this different from the return-by-value performed by the compiler?

Is this copy function being used in a different commit? (I couldn't find any appropriate usages of this `copy` function in this commit.)

Reply:
It is used in later commits. It copies so that an lvalue reference can be passed to a function expecting an rvalue reference.

Reply:
okay